c++/src/capnp/compat/http-over-capnp-test.c++

```diff
@@ -648,5 +648,194 @@ KJ_TEST("HttpService isn't destroyed while call outstanding") {
+      fulfiller->fulfill(
+        kj::HttpClient::ConnectRequest::Status(
+          statusCode,
+          kj::str(statusText),
+          kj::heap(headers.clone()),
+          nullptr
+        )
+      );
+    }
+
+    kj::Own<kj::AsyncOutputStream> reject(
+        uint statusCode,
+        kj::StringPtr statusText,
+        const kj::HttpHeaders& headers,
+        kj::Maybe<uint64_t> expectedBodySize) override {
+      KJ_UNREACHABLE;
+    }
+  };
+
+  auto clientPipe = kj::newTwoWayPipe();
+  auto paf = kj::newPromiseAndFulfiller<kj::HttpClient::ConnectRequest::Status>();
+  ResponseImpl response(kj::mv(paf.fulfiller));
+
+  auto promise = frontCapnpHttpService->connect(
+      "https://example.org"_kj, kj::HttpHeaders(*table), *clientPipe.ends[0],
+      response).attach(kj::mv(clientPipe.ends[0]));
+
+  paf.promise.then(
+      [io = kj::mv(clientPipe.ends[1])](auto status) mutable {
+    KJ_ASSERT(status.statusCode == 200);
+    KJ_ASSERT(status.statusText == "OK"_kj);
+
+    auto buf = kj::heapArray<char>(4);
+    return io->tryRead(buf.begin(), 4, 4).then(
+        [buf = kj::mv(buf), io = kj::mv(io)](size_t count) mutable {
+      KJ_ASSERT(count == 4, "Expecting the stream to read 4 chars.");
+      return io->tryRead(buf.begin(), 1, 1).then(
+          [buf = kj::mv(buf)](size_t count) mutable {
+        KJ_ASSERT(count == 0, "Expecting the stream to get disconnected.");
+      }).attach(kj::mv(io));
+    });
+  }).wait(waitScope);
+
+  listenTask.wait(waitScope);
+}
+
+
+KJ_TEST("HTTP-over-Cap'n-Proto Connect Reject") {
+  kj::EventLoop eventLoop;
+  kj::WaitScope waitScope(eventLoop);
+
+  auto pipe = kj::newTwoWayPipe();
+
+  kj::TimerImpl timer(kj::origin<kj::TimePoint>());
+
+  ByteStreamFactory streamFactory;
+  kj::HttpHeaderTable::Builder tableBuilder;
+  HttpOverCapnpFactory factory(streamFactory, tableBuilder);
+  kj::Own<kj::HttpHeaderTable> table = tableBuilder.build();
+  ConnectRejectService service(*table);
+  kj::HttpServer server(timer, *table, service);
+
+  auto listenTask = server.listenHttp(kj::mv(pipe.ends[0]));
+
+  auto client = newHttpClient(*table, *pipe.ends[1]);
+
+  capnp::HttpService::Client httpService = factory.kjToCapnp(newHttpService(*client));
+  auto frontCapnpHttpService = factory.capnpToKj(httpService);
+
+  struct ResponseImpl final: public kj::HttpService::ConnectResponse {
+    kj::Own<kj::PromiseFulfiller<kj::Own<kj::AsyncInputStream>>> fulfiller;
+    ResponseImpl(kj::Own<kj::PromiseFulfiller<kj::Own<kj::AsyncInputStream>>> fulfiller)
+      : fulfiller(kj::mv(fulfiller)) {}
+    void accept(uint statusCode, kj::StringPtr statusText, const kj::HttpHeaders& headers) override {
+      KJ_UNREACHABLE;
+    }
+
+    kj::Own<kj::AsyncOutputStream> reject(
+        uint statusCode,
+        kj::StringPtr statusText,
+        const kj::HttpHeaders& headers,
+        kj::Maybe<uint64_t> expectedBodySize) override {
+      KJ_ASSERT(statusCode == 500);
```

Comment:
Two things I noticed while writing this test:

* A KJ_ASSERT failure here does not show up in test logs (all I get is "failed: Nothing to wait for; this thread would hang forever." which isn't ideal).
* Hopefully not a blocker for this PR but the `expectedBodySize` appears to be getting eaten by https://github.com/capnproto/capnproto/blob/master/c%2B%2B/src/kj/compat/http.c%2B%2B#L6510

Reply:
Hmm, the first point, about KJ_ASSERT, seems worth understanding. :(

Regarding `expectedBodySize`, I wonder if the linked line needs to call `errorBody->tryGetLength()` and pass its result to `reject()`? Though for this particular `expectedBodySize` (line 810 above), we should probably pass it to the `kj::newOneWayPipe()` on line 813 below.

Reply:
KJ_ASSERT throws an exception. The exception could be caught somewhere, or (more likely) could be held up in a promise that nothing is waiting on, or is one branch of a `kj::joinPromises()` and the other branch is hanging.

> Hopefully not a blocker for this PR but the expectedBodySize appears to be getting eaten by

Looks like it should be `errorBody->tryGetLength()`. Can you go ahead and fix that before we forget?